Refactor Upgrade file for 2.0 (#17448)
Sylius/Sylius: UPGRADE-2.0.md

# UPGRADE FROM `1.14` TO `2.0`

> ## To start of

Reviewer: ```suggestion
## To start off
```

-   JMS\SerializerBundle\JMSSerializerBundle::class => ['all' => true],
-   FOS\RestBundle\FOSRestBundle::class => ['all' => true],
-   ApiPlatform\Core\Bridge\Symfony\Bundle\ApiPlatformBundle::class => ['all' => true],
-   SyliusLabs\Polyfill\Symfony\Security\Bundle\SyliusLabsPolyfillSymfonySecurityBundle::class => ['all' => true],
+   ApiPlatform\Symfony\Bundle\ApiPlatformBundle::class => ['all' => true],
+   Sylius\TwigHooks\SyliusTwigHooksBundle::class => ['all' => true],
+   Symfony\UX\TwigComponent\TwigComponentBundle::class => ['all' => true],
+   Symfony\UX\StimulusBundle\StimulusBundle::class => ['all' => true],
+   Symfony\UX\LiveComponent\LiveComponentBundle::class => ['all' => true],
+   Symfony\UX\Autocomplete\AutocompleteBundle::class => ['all' => true],
];

* New Symfony/Messenger transports for handling payment requests have been added. 
Therefore, you need to add the following configuration to your .env file:

```diff
###> symfony/messenger ###
...
+ SYLIUS_MESSENGER_TRANSPORT_PAYMENT_REQUEST_DSN=sync://
+ SYLIUS_MESSENGER_TRANSPORT_PAYMENT_REQUEST_FAILED_DSN=sync://
###< symfony/messenger ###
```

> ### Migrations

Author: I'm not sure where to put that part, it might be also at the ned 🤔

Doctrine migrations have been regenerated, meaning all previous migration files have been removed and their content is now in a single migration file.
To apply the new migration and get rid of the old entries run migrations as usual:

```bash
    bin/console doctrine:migrations:migrate --no-interaction
```

> ### PHP 8.2 support

Reviewer 2: ```suggestion
### PHP support
```

Sylius 2.0 comes with a bump of minimum PHP version to 8.2. 

### Symfony support

The support of Symfony 5.4 has been dropped.
Sylius 2.0 supports both Symfony 6.4 and 7.1

## Main Update

Once you’ve applied these initial changes, your app should be able to run. However, depending on the customizations
you’ve made, you may need to make some additional adjustments. Carefully review the following changes and apply them to
your app as necessary.

### Dependencies

#### Replaced:

* The `swiftmailer/swiftmailer` dependency has been removed. Use `symfony/mailer` instead.

#### Moved:

> * The `sylius/theme-bundle` dependency has been moved from CoreBundle to ShopBundle, so it will no longer be installed

Reviewer: ```suggestion
* The `sylius/theme-bundle` dependency has been moved from CoreBundle to ShopBundle.
```

        * `sylius-labs/polyfill-symfony-event-dispatcher`
        * `sylius-labs/polyfill-symfony-framework-bundle`

#### Optional

> * Removed from required and made suggested, install them in your application, if you still want to use Winzou State Machine:

Reviewer: ```suggestion
* Winzou State Machine repositories have been moved to the suggested section of composer.json,
if you still want to use them, install the following repositories:
```

        )
    ```

   `Sylius\Component\Core\OrderProcessing\OrderPaymentProcessor`
    ```diff
        public function __construct(
            private OrderPaymentProviderInterface $orderPaymentProvider,
    -       private string $targetState = PaymentInterface::STATE_CART,
            private OrderPaymentsRemoverInterface $orderPaymentsRemover,
            private array $unprocessableOrderStates,
    +       private string $targetState = PaymentInterface::STATE_CART,
        )
    ```

1. Change across the codebase:
    ```diff
    -   private StateMachineInterface $stateMachineFactory,
    +   private StateMachineInterface $stateMachine,
    ```

### Configuration Changes

#### Messenger

- Removed `sylius_default.bus` and `sylius_event.bus` configurations.  
  Use `sylius.command_bus` and `sylius.event_bus` for commands and events, respectively.

#### Sylius State Machine Abstraction

- Changed `sylius_state_machine_abstraction.default_adapter` from `winzou_state_machine` to `symfony_workflow`.

#### Resource

- Removed configuration nodes for resource options `sylius_*.resources.**.options`, such as `sylius_addressing.resources.address.options`.

> - Removed `sylius_inventory.checker` configuartion node has been removed

Reviewer: ```suggestion
- Removed `sylius_inventory.checker` configuration node.
```

#### Autoconfiguration

- Removed `sylius_core.autoconfigure_with_attributes` and `sylius_order.autoconfigure_with_attributes`.  
  Use the following attributes instead of interfaces for autoconfiguration:
    - `Sylius\Bundle\OrderBundle\Attribute\AsCartContext`
    - `Sylius\Bundle\OrderBundle\Attribute\AsOrderProcessor`

#### User Resetting Pin

- Removed `sylius_user.resources.{name}.user.resetting.pin` configuration parameter.  
  Also removed:
    - `reset_password_pin` email
    - `Sylius\Bundle\UserBundle\Controller\UserController::requestPasswordResetPinAction` method
    - Related services:
        - `sylius.{user_type}_user.pin_generator.password_reset`

>         - `sylius.{user_type}_user.pin_uniqueness_checker.password_reset`

Reviewer: ```suggestion
  Due to that the related logic has also been removed, this includes:
    - `reset_password_pin` email
    - `Sylius\Bundle\UserBundle\Controller\UserController::requestPasswordResetPinAction` method
    - `sylius.{user_type}_user.pin_generator.password_reset` services
    - `sylius.{user_type}_user.pin_uniqueness_checker.password_reset` services
```

#### Parameter Removals and Renaming

- Removed:
    - `sylius.mongodb_odm.repository.class`
    - `sylius.phpcr_odm.repository.class`

- Renamed:
    - `sylius.message.admin_user_create.validation_groups`  
      → `sylius_admin.command_handler.create_admin_user.validation_groups`

- Removed:

>     - `sylius.mailer.templates`

Reviewer: ```suggestion
- Removed:
    - `sylius.mongodb_odm.repository.class`
    - `sylius.phpcr_odm.repository.class`
    - `sylius.mailer.templates`

- Renamed:
    - `sylius.message.admin_user_create.validation_groups` to `sylius_admin.command_handler.create_admin_user.validation_groups`
```

#### File Relocations

- Winzou state machine configuration:
    - Moved `state_machine.yml` from `@SyliusPaymentBundle/Resources/config/app` to `@SyliusPaymentBundle/Resources/config/app/state_machine`.
    - Renamed to `sylius_payment.yaml`.

- Symfony workflow configuration:
    - Moved `state_machine.yaml` from `@SyliusPaymentBundle/Resources/config/workflow` to `@SyliusPaymentBundle/Resources/config/app/workflow`.

>     - Renamed to `sylius_payment.yaml`.

Reviewer: ```suggestion
- The state machine configurations of `PaymentBundle` have been moved and renamed:

Winzou state machine:
```diff
- `@SyliusPaymentBundle/Resources/config/app/state_machine.yml`
+ `@SyliusPaymentBundle/Resources/config/app/state_machine/sylius_payment.yaml`
\```

Symfony workflow:
```diff
- `@SyliusPaymentBundle/Resources/config/workflow/state_machine.yaml`
+ `@SyliusPaymentBundle/Resources/config/app/workflow/sylius_payment.yaml`
\```
```

-                   type: entities
+                   type: entity
                    label: sylius.ui.channel
                    form_options:
                        class: "%sylius.model.channel.class%"
                    options:
-                       field: product.channels.id
+                       fields: [product.channels.id]
```

#### Security

* API firewalls have been renamed and user checkers have been configured on
  them with `security.user_checker.chain.<firewall>` services:

    ```diff
    security:
        firewalls:
            admin:
                ...
    +           user_checker: security.user_checker.chain.admin
    -       new_api_admin_user:
    +       api_admin:
                ...
    +           user_checker: security.user_checker.chain.api_admin
    -       new_api_shop_user:
    +       api_shop:
                ...
    +           user_checker: security.user_checker.chain.api_shop
            shop:
                ...
    +           user_checker: security.user_checker.chain.shop
    ```

#### Password Encoder & Salt

The encoder and salt has been removed from the User entities. It will use the password hasher configured on Symfony

> security configuration.

Reviewer: ```suggestion
The encoder and salt have been removed from the User entities. The password hasher configured on Symfony security configuration is in use instead.
```

This "encoder" is configured via
the [Symfony security password hasher](https://symfony.com/doc/current/security/passwords.html#configuring-a-password-hasher).

> You may have already something like that configuration bellow.

Reviewer: ```suggestion
For example:
```

```yaml
# config/packages/security.yaml
security:
    # ...

    password_hashers:
        Sylius\Component\User\Model\UserInterface: argon2i
```

> Check if you have an encoder configured in the `sylius_user` package configuration.

Reviewer: ```suggestion
Also, check if you have an encoder configured in the `sylius_user` package configuration.
```

```yaml
sylius_user:
    # ...

    encoder: plaintext # Remove this line

    # ...
    resources:
        oauth:
            user:
                encoder: false # Remove this line
                classes: Sylius\Component\User\Model\UserOAuth
```

Check your user hashed passwords in your production database.

> In modern Symfony projects, the hasher name is stored on the password.

Reviewer: ```suggestion
In modern Symfony projects, the hasher name is stored along with the password.
```

    * `sylius_admin_partial_taxon_tree`
    * `sylius_admin_render_attribute_forms`
    * `sylius_shop_ajax_cart_add_item`
    * `sylius_shop_ajax_cart_item_remove`
    * `sylius_shop_ajax_user_check_action`
    * `sylius_shop_partial_cart_summary`
    * `sylius_shop_partial_cart_add_item`
    * `sylius_shop_partial_channel_menu_taxon_index`
    * `sylius_shop_partial_product_association_show`
    * `sylius_shop_partial_product_index_latest`
    * `sylius_shop_partial_product_review_latest`
    * `sylius_shop_partial_product_show_by_slug`
    * `sylius_shop_partial_taxon_index_by_code`
    * `sylius_shop_partial_taxon_show_by_slug`

### Services, classes and interfaces changes

#### Container

In Sylius 2.0, we have changed the visibility of services to `private` by default. This change enhances the performance and maintainability of the application and also follows Symfony's best practices for service encapsulation.

**Exceptions:**
- Services required by Symfony to be `public` (e.g., controllers, event listeners) remain public.
- Services used in `ResourceController` must be `public` as they are accessed directly from the container.

#### Classes and Interfaces

1. Removed

> 

Reviewer: ```suggestion
```

| `sylius.twig.extension.form_test_attribute_array`                                  | `sylius_twig_extra.twig.extension.test_form_attribute`                   |
| `sylius.twig.extension.form_test_attribute_name`                                   | `sylius_twig_extra.twig.extension.test_html_attribute`                   |
| `sylius.twig.extension.sort_by`                                                    | `sylius_twig_extra.twig.extension.sort_by`                               |
| `Sylius\Bundle\UiBundle\Twig\RouteExistsExtension`                                 | `sylius_twig_extra.twig.extension.route_exists`                          |
| `sylius.form_registry.payum_gateway_config`                                        | `sylius.form_registry.payment_gateway_config` (moved to `PaymentBundle`) |
| `Sylius\Bundle\PayumBundle\Validator\GatewayFactoryExistsValidator`                | `sylius.validator.gateway_factory_exists`                                |
| `Sylius\Bundle\PayumBundle\Validator\GroupsGenerator\GatewayConfigGroupsGenerator` | `sylius.validator.groups_generator.gateway_config`                       |

> 

Reviewer: ```suggestion
```

    - `Sylius\Bundle\AddressingBundle\Doctrine\ORM\ZoneMemberRepository`

- **Attribute**:
    - `Sylius\Bundle\AttributeBundle\Doctrine\ORM\AttributeRepository`
    - `Sylius\Bundle\AttributeBundle\Doctrine\ORM\AttributeTranslationRepository`
    - `Sylius\Bundle\AttributeBundle\Doctrine\ORM\AttributeValueRepository`

- **Product**:
    - `Sylius\Bundle\ProductBundle\Doctrine\ORM\ProductAssociationRepository`
    - `Sylius\Bundle\ProductBundle\Doctrine\ORM\ProductAssociationTypeTranslationRepository`
    - `Sylius\Bundle\ProductBundle\Doctrine\ORM\ProductAttributeRepository`
    - `Sylius\Bundle\ProductBundle\Doctrine\ORM\ProductAttributeTranslationRepository`
    - `Sylius\Bundle\ProductBundle\Doctrine\ORM\ProductOptionTranslationRepository`
    - `Sylius\Bundle\ProductBundle\Doctrine\ORM\ProductOptionValueRepository`
    - `Sylius\Bundle\ProductBundle\Doctrine\ORM\ProductOptionValueTranslationRepository`
    - `Sylius\Bundle\ProductBundle\Doctrine\ORM\ProductTranslationRepository`
    - `Sylius\Bundle\ProductBundle\Doctrine\ORM\ProductVariantTranslationRepository`

- **Currency**:
    - `Sylius\Bundle\CurrencyBundle\Doctrine\ORM\CurrencyRepository`

> ##### Repository Namespace Changes

Reviewer: ```suggestion
##### Repository Inheritance Changes
```
at least based on the content

- **Addressing**:
    - `Sylius\Bundle\CoreBundle\Doctrine\ORM\AddressRepository` now extends `Sylius\Bundle\AddressingBundle\Doctrine\ORM\AddressRepository`.
    - `Sylius\Component\Core\Repository\AddressRepositoryInterface` now implements `Sylius\Component\Addressing\Repository\AddressRepositoryInterface`.

- **Attribute**:
    - `Sylius\Bundle\CoreBundle\Doctrine\ORM\AttributeRepository` now extends `Sylius\Bundle\AttributeBundle\Doctrine\ORM\AttributeRepository`.
    - `Sylius\Component\Product\Repository\ProductAttributeValueRepositoryInterface` now implements `Sylius\Component\Attribute\Repository\AttributeValueRepositoryInterface`.

- **Product**:
    - `Sylius\Bundle\CoreBundle\Doctrine\ORM\ProductAssociationRepository` now extends `Sylius\Bundle\ProductBundle\Doctrine\ORM\ProductAssociationRepository`.

#### User Model Updates

- Removed fields and corresponding methods for:
    - `locked`
    - `expiresAt`
    - `credentialsExpireAt`

> - These changes affect both the ShopUser and AdminUser models, as well as the relevant columns in the database tables.

Reviewer: ```suggestion
- These changes affect the ShopUser and AdminUser models, and any custom user type extending the `Sylius\Component\User\Model\User` model, as well as the relevant columns in the database tables.
```

> #### Aliases

Reviewer: ```suggestion
#### Service Aliases
```

### Architecture changes

* The `Normalizer` and `Denormalizer` in `ApiBundle` have been reorganized into subdirectories, moving from `Sylius\Bundle\ApiBundle\Serializer`
  to `Sylius\Bundle\ApiBundle\Serializer\Normalizer` and `Sylius\Bundle\ApiBundle\Serializer\Denormalizer`.

* The `SerializerContextBuilder` classes in `ApiBundle` have been moved to the `Sylius\Bundle\ApiBundle\Serializer\ContextBuilder` subdirectory.

* The `Filter`, `QueryCollectionExtension`, and `QueryItemExtension` classes in `ApiBundle` have been reorganized into separate sections (`admin`, `shop`, `common`)
  based on their usage context, and grouped by resources.

* `AdminBundle` now contains base form types for every resource. 
Use these as an extension point for admin customizations instead the ones from `CoreBundle` or other minor bundles.
Similarly, use form types from `ShopBundle` for the Shop context.

### Payment method gateways

* Stripe and Paypal Express Checkout gateways have been removed.

> The only one included in core is `Offline`. Gateways are now available as separate plugins.

Reviewer: ```suggestion
The only remaining by default gateway in core is `offline`. Use sylius plugins for the gateways of your choosing.
```

### Theming

* Channel's `themeName` form field existence is made optional and depends on `ShopBundle` presence.
* The `Sylius\Bundle\CoreBundle\Theme\ChannelBasedThemeContext` has been moved to 
  the `Sylius\Bundle\ShopBundle\Theme\ChannelBasedThemeContext`.

> ### Testing Suite

Reviewer: This section can be moved to the bottom of the file

* The recommended Node.js versions are 20 or 22, as support for version 18 has been dropped.
* `use_webpack` option was removed from the `sylius_ui` configuration, and the Webpack has become the only module
  bundler provided by Sylius.
* `use_webpack` twig global variable was removed. Webpack is always used now, and there is no need to check for it.
* Image sizes have been simplified and standardized for both the Admin and Shop Bundle.
* Some Twig extension services have been moved from the UiBundle to the new Twig Extra package

#### KNP Menu

* Aliases for the `knp_menu.menu_builder` tags introduced in Sylius 1.14 are now the only valid menu builder tags in
  Sylius 2.0:

| Old Alias             | New Alias             |
|-----------------------|-----------------------|
| **AdminBundle**       |                       |
| `sylius.admin.main`   | `sylius_admin.main`   |
| **ShopBundle**        |                       |
| `sylius.shop.account` | `sylius_shop.account` |

#### Transition from SemanticUI to Bootstrap

> - All `.ui` prefixed classes have been replaced with Bootstrap classes.

Reviewer: ```suggestion
- All CSS classes of SemanticUI have been replaced with Bootstrap classes.
```
A bit of pedanticism, but not all semantic classes were prefixed.

- JavaScript components relying on Semantic UI have been rewritten to utilize Bootstrap's JavaScript plugins.
- Customized CSS has been replaced by Bootstrap's utility classes.

> #### Abandoning jQuery

Reviewer: ```suggestion
#### Removal of jQuery
```

Most of the existing JavaScript has been replaced by SymfonyUX with Stimulus, which includes live components.
This change led to the removal of jQuery and a significant reduction of custom JavaScript in the project. 
Check out the documentation for more information [here](https://ux.symfony.com/).

#### Abandoning partial routes

All partial routes rendered in templates have been removed and replaced by rendering Twig components.

#### Sylius Twig Hooks

> Twig Hooks are a robust and powerful alternative to the Sonata Block Events and the old Sylius Template Events systems.

Reviewer: Maybe link `Twig Hooks` with the repo URL?

> ##### Abandoning Sonata Blocks

Reviewer: ```suggestion
##### Removal of Sonata Blocks
```

> Sonata Blocks have been fully replaced with Twig Hooks.

Reviewer: ```suggestion
Sonata Blocks have been fully removed as they were not actively maintained/supported for a long time.
```
The previous fit more w/ template events

> ##### Evolving Sylius Template Events to Twig Hooks

Reviewer: Maybe add a note regarding the templates directory structure as well?
Since we have a pretty much 1:1 regarding hook name:template location, would be nice to flesh it out.
Probably will also help with the upgrade when people know more or less where the parts they need to change are located/how the appropriate hook can be named.

+                       fields: [product.channels.id]
```

#### Security

* API firewalls have been renamed and user checkers have been configured on
  them with `security.user_checker.chain.<firewall>` services:

    ```diff
    security:
        firewalls:
            admin:
                ...
    +           user_checker: security.user_checker.chain.admin
    -       new_api_admin_user:
    +       api_admin:
                ...
    +           user_checker: security.user_checker.chain.api_admin
    -       new_api_shop_user:
    +       api_shop:
                ...
    +           user_checker: security.user_checker.chain.api_shop
            shop:
                ...
    +           user_checker: security.user_checker.chain.shop

>     ```

Reviewer 2: This section is duplicated:
```suggestion
```

    * `sylius_admin_partial_address_log_entry_index`
    * `sylius_admin_partial_catalog_promotion_show`
    * `sylius_admin_partial_channel_index`
    * `sylius_admin_partial_customer_latest`
    * `sylius_admin_partial_customer_show`
    * `sylius_admin_partial_order_latest`
    * `sylius_admin_partial_order_latest_in_channel`
    * `sylius_admin_partial_product_show`
    * `sylius_admin_partial_promotion_show`
    * `sylius_admin_partial_taxon_show`
    * `sylius_admin_partial_taxon_tree`
    * `sylius_admin_render_attribute_forms`
    * `sylius_shop_ajax_cart_add_item`
    * `sylius_shop_ajax_cart_item_remove`
    * `sylius_shop_ajax_user_check_action`
    * `sylius_shop_partial_cart_summary`
    * `sylius_shop_partial_cart_add_item`
    * `sylius_shop_partial_channel_menu_taxon_index`
    * `sylius_shop_partial_product_association_show`
    * `sylius_shop_partial_product_index_latest`
    * `sylius_shop_partial_product_review_latest`
    * `sylius_shop_partial_product_show_by_slug`
    * `sylius_shop_partial_taxon_index_by_code`
    * `sylius_shop_partial_taxon_show_by_slug`

### Services, classes and interfaces changes

#### Container

> In Sylius 2.0, we have changed the visibility of services to `private` by default. This change enhances the performance and maintainability of the application and also follows Symfony's best practices for service encapsulation.

Reviewer 2: This line could be split into several ones

                    /show.yaml
```

- **Templates**

> The structure of the template directories has also been improved. Previously, templates were organized in a less intuitive way, with some structures inconsistent with the template event definitions:

Reviewer 2: Split this line into several ones, the same applies to some other points
